First class callable causes crash when method name is a variable

It seems Psalm crashes when method name for the first class callable comes from a variable, e.g. `TestClass::$method(...)`.

Reproduction here: https://psalm.dev/r/5cc345f634

Sample stack trace for the error:

```
λ composer psalm
> psalm
Target PHP version: 8.1 (inferred from current PHP version)
Scanning files...
Analyzing files...

░░EE░░░░E░E░░░░░░░░░░EEE░░░░░░░░░░EE░░Uncaught AssertionError: assert(!$this->isFirstClassCallable()) in D:\Projects\type-kit\vendor\nikic\php-parser\lib\PhpParser\Node\Expr\CallLike.php:36
Stack trace:
#0 D:\Projects\type-kit\vendor\nikic\php-parser\lib\PhpParser\Node\Expr\CallLike.php(36): assert(false, 'assert(!$this->...')
#1 D:\Projects\type-kit\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Expression\Call\StaticMethod\AtomicStaticCallAnalyzer.php(229): PhpParser\Node\Expr\CallLike->getArgs()
#2 D:\Projects\type-kit\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\Expression\Call\StaticCallAnalyzer.php(215): Psalm\Internal\Analyzer\Statements\Expression\Call\StaticMethod\AtomicStaticCallAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\StaticCall), Object(Psalm\Context), Object(Psalm\Type\Atomic\TNamedObject), false, false, false, false)
#3 D:\Projects\type-kit\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer.php(190): Psalm\Internal\Analyzer\Statements\Expression\Call\StaticCallAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\StaticCall), Object(Psalm\Context))
#4 D:\Projects\type-kit\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer.php(78): Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::handleExpression(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\StaticCall), Object(Psalm\Context), false, NULL, false)
#5 D:\Projects\type-kit\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\Statements\ReturnAnalyzer.php(154): Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\StaticCall), Object(Psalm\Context))
#6 D:\Projects\type-kit\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(543): Psalm\Internal\Analyzer\Statements\ReturnAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\Return_), Object(Psalm\Context))
#7 D:\Projects\type-kit\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\StatementsAnalyzer.php(206): Psalm\Internal\Analyzer\StatementsAnalyzer::analyzeStatement(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\Return_), Object(Psalm\Context), Object(Psalm\Context))
#8 D:\Projects\type-kit\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\FunctionLikeAnalyzer.php(467): Psalm\Internal\Analyzer\StatementsAnalyzer->analyze(Array, Object(Psalm\Context), Object(Psalm\Context), true)
#9 D:\Projects\type-kit\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\ClassAnalyzer.php(1788): Psalm\Internal\Analyzer\FunctionLikeAnalyzer->analyze(Object(Psalm\Context), Object(Psalm\Internal\Provider\NodeDataProvider), Object(Psalm\Context))
#10 D:\Projects\type-kit\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\ClassAnalyzer.php(425): Psalm\Internal\Analyzer\ClassAnalyzer->analyzeClassMethod(Object(PhpParser\Node\Stmt\ClassMethod), Object(Psalm\Storage\ClassLikeStorage), Object(Psalm\Internal\Analyzer\ClassAnalyzer), Object(Psalm\Context), Object(Psalm\Context))
#11 D:\Projects\type-kit\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\FileAnalyzer.php(229): Psalm\Internal\Analyzer\ClassAnalyzer->analyze(Object(Psalm\Context), Object(Psalm\Context))
#12 D:\Projects\type-kit\vendor\vimeo\psalm\src\Psalm\Internal\Codebase\Analyzer.php(362): Psalm\Internal\Analyzer\FileAnalyzer->analyze()
#13 D:\Projects\type-kit\vendor\vimeo\psalm\src\Psalm\Internal\Codebase\Analyzer.php(619): Psalm\Internal\Codebase\Analyzer->Psalm\Internal\Codebase\{closure}(38, 'D:\\Projects\\typ...')
#14 D:\Projects\type-kit\vendor\vimeo\psalm\src\Psalm\Internal\Codebase\Analyzer.php(291): Psalm\Internal\Codebase\Analyzer->doAnalysis(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 1)
#15 D:\Projects\type-kit\vendor\vimeo\psalm\src\Psalm\Internal\Analyzer\ProjectAnalyzer.php(685): Psalm\Internal\Codebase\Analyzer->analyzeFiles(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 1, false, true)
#16 D:\Projects\type-kit\vendor\vimeo\psalm\src\Psalm\Internal\Cli\Psalm.php(373): Psalm\Internal\Analyzer\ProjectAnalyzer->check('D:\\Projects\\typ...', true)
#17 D:\Projects\type-kit\vendor\vimeo\psalm\psalm(7): Psalm\Internal\Cli\Psalm::run(Array)
#18 D:\Projects\type-kit\vendor\bin\psalm(107): include('D:\\Projects\\typ...')
#19 {main}
(Psalm 4.18.1@dda05fa913f4dc6eb3386f2f7ce5a45d37a71bcb crashed due to an uncaught Throwable)
Script psalm handling the psalm event returned with error code 1
```

--- a/src/Psalm/Internal/Analyzer/Statements/Expression/Call/Method/AtomicMethodCallAnalyzer.php
+++ b/src/Psalm/Internal/Analyzer/Statements/Expression/Call/Method/AtomicMethodCallAnalyzer.php
@@ -27,6 +27,7 @@ use Psalm\StatementsSource;
 use Psalm\Storage\ClassLikeStorage;
 use Psalm\Type;
 use Psalm\Type\Atomic;
+use Psalm\Type\Atomic\TClosure;
 use Psalm\Type\Atomic\TEmpty;
 use Psalm\Type\Atomic\TEmptyMixed;
 use Psalm\Type\Atomic\TFalse;
@@ -186,6 +187,41 @@ class AtomicMethodCallAnalyzer extends CallAnalyzer
                     strtolower($fq_class_name) . '::',
                     $context->calling_method_id ?: $statements_analyzer->getFileName()
                 );
+            }
+
+            if ($stmt->isFirstClassCallable()) {
+                $return_type_candidate = null;
+                $method_name_type = $statements_analyzer->node_data->getType($stmt->name);
+                if ($method_name_type && $method_name_type->isSingleStringLiteral()) {
+                    $method_identifier = new MethodIdentifier(
+                        $fq_class_name,
+                        strtolower($method_name_type->getSingleStringLiteral()->value)
+                    );
+                    //the call to methodExists will register that the method was called from somewhere
+                    if ($codebase->methods->methodExists(
+                        $method_identifier,
+                        $context->calling_method_id,
+                        null,
+                        $statements_analyzer,
+                        $statements_analyzer->getFilePath(),
+                        true,
+                        $context->insideUse()
+                    )) {
+                        $method_storage = $codebase->methods->getStorage($method_identifier);
+
+                        $return_type_candidate = new Union([new TClosure(
+                            'Closure',
+                            $method_storage->params,
+                            $method_storage->return_type,
+                            $method_storage->pure
+                        )]);
+                    }
+                }
+
+                $statements_analyzer->node_data->setType($stmt, $return_type_candidate ?? Type::getClosure());
+
+
+                return;
             }
 
             ArgumentsAnalyzer::analyze(

--- a/src/Psalm/Internal/Analyzer/Statements/Expression/Call/StaticMethod/AtomicStaticCallAnalyzer.php
+++ b/src/Psalm/Internal/Analyzer/Statements/Expression/Call/StaticMethod/AtomicStaticCallAnalyzer.php
@@ -227,30 +227,32 @@ class AtomicStaticCallAnalyzer
 
             if ($stmt->isFirstClassCallable()) {
                 $return_type_candidate = null;
-                $method_name_type = $statements_analyzer->node_data->getType($stmt->name);
-                if ($method_name_type && $method_name_type->isSingleStringLiteral()) {
-                    $method_identifier = new MethodIdentifier(
-                        $fq_class_name,
-                        strtolower($method_name_type->getSingleStringLiteral()->value)
-                    );
-                    //the call to methodExists will register that the method was called from somewhere
-                    if ($codebase->methods->methodExists(
-                        $method_identifier,
-                        $context->calling_method_id,
-                        null,
-                        $statements_analyzer,
-                        $statements_analyzer->getFilePath(),
-                        true,
-                        $context->insideUse()
-                    )) {
-                        $method_storage = $codebase->methods->getStorage($method_identifier);
+                if (!$stmt->name instanceof PhpParser\Node\Identifier) {
+                    $method_name_type = $statements_analyzer->node_data->getType($stmt->name);
+                    if ($method_name_type && $method_name_type->isSingleStringLiteral()) {
+                        $method_identifier = new MethodIdentifier(
+                            $fq_class_name,
+                            strtolower($method_name_type->getSingleStringLiteral()->value)
+                        );
+                        //the call to methodExists will register that the method was called from somewhere
+                        if ($codebase->methods->methodExists(
+                            $method_identifier,
+                            $context->calling_method_id,
+                            null,
+                            $statements_analyzer,
+                            $statements_analyzer->getFilePath(),
+                            true,
+                            $context->insideUse()
+                        )) {
+                            $method_storage = $codebase->methods->getStorage($method_identifier);
 
-                        $return_type_candidate = new Union([new TClosure(
-                            'Closure',
-                            $method_storage->params,
-                            $method_storage->return_type,
-                            $method_storage->pure
-                        )]);
+                            $return_type_candidate = new Union([new TClosure(
+                                'Closure',
+                                $method_storage->params,
+                                $method_storage->return_type,
+                                $method_storage->pure
+                            )]);
+                        }
                     }
                 }
 

--- a/src/Psalm/Internal/Analyzer/Statements/Expression/Call/StaticMethod/AtomicStaticCallAnalyzer.php
+++ b/src/Psalm/Internal/Analyzer/Statements/Expression/Call/StaticMethod/AtomicStaticCallAnalyzer.php
@@ -225,6 +225,40 @@ class AtomicStaticCallAnalyzer
                 );
             }
 
+            if ($stmt->isFirstClassCallable()) {
+                $return_type_candidate = null;
+                $method_name_type = $statements_analyzer->node_data->getType($stmt->name);
+                if ($method_name_type && $method_name_type->isSingleStringLiteral()) {
+                    $method_identifier = new MethodIdentifier(
+                        $fq_class_name,
+                        strtolower($method_name_type->getSingleStringLiteral()->value)
+                    );
+                    //the call to methodExists will register that the method was called from somewhere
+                    if ($codebase->methods->methodExists(
+                        $method_identifier,
+                        $context->calling_method_id,
+                        null,
+                        $statements_analyzer,
+                        $statements_analyzer->getFilePath(),
+                        true,
+                        $context->insideUse()
+                    )) {
+                        $method_storage = $codebase->methods->getStorage($method_identifier);
+
+                        $return_type_candidate = new Union([new TClosure(
+                            'Closure',
+                            $method_storage->params,
+                            $method_storage->return_type,
+                            $method_storage->pure
+                        )]);
+                    }
+                }
+
+                $statements_analyzer->node_data->setType($stmt, $return_type_candidate ?? Type::getClosure());
+
+                return;
+            }
+
             if (ArgumentsAnalyzer::analyze(
                 $statements_analyzer,
                 $stmt->getArgs(),

--- a/tests/ClosureTest.php
+++ b/tests/ClosureTest.php
@@ -609,6 +609,27 @@ class ClosureTest extends TestCase
                 [],
                 '8.1'
             ],
+            'FirstClassCallable:InstanceMethod:Expr' => [
+                '<?php
+                    class Test {
+                        public function __construct(private readonly string $string) {
+                        }
+
+                        public function length(): int {
+                            return strlen($this->string);
+                        }
+                    }
+                    $test = new Test("test");
+                    $method_name = "length";
+                    $closure = $test->$method_name(...);
+                    $length = $closure();
+                ',
+                'assertions' => [
+                    '$length' => 'int',
+                ],
+                [],
+                '8.1'
+            ],
             'FirstClassCallable:InstanceMethod:BuiltIn' => [
                 '<?php
                     $queue = new \SplQueue;
@@ -629,6 +650,23 @@ class ClosureTest extends TestCase
                         }
                     }
                     $closure = Test::length(...);
+                    $length = $closure("test");
+                ',
+                'assertions' => [
+                    '$length' => 'int',
+                ],
+                [],
+                '8.1'
+            ],
+            'FirstClassCallable:StaticMethod:Expr' => [
+                '<?php
+                    class Test {
+                        public static function length(string $param): int {
+                            return strlen($param);
+                        }
+                    }
+                    $method_name = "length";
+                    $closure = Test::$method_name(...);
                     $length = $closure("test");
                 ',
                 'assertions' => [

